Create/Edit tools called with empty arguments when using Google provider

## Description

When using ProbeAgent with Google (Gemini) provider and `allowEdit: true`, the `create` and `edit` tools are called with empty arguments, causing them to fail with validation errors.

## Environment

- Probe version: 0.6.0-rc196
- Provider: Google (Gemini)
- Node version: v24.7.0

## Steps to Reproduce

1. Configure ProbeAgent with:
   - `provider: google`
   - `allowEdit: true`
   - A prompt that instructs the AI to create a file

2. The AI attempts to use the `create` tool but passes no arguments

## Expected Behavior

The AI should call the create tool with proper XML parameters:
```xml
<create>
<file_path>./my-file.yaml</file_path>
<content>file content here</content>
</create>
```

## Actual Behavior

Debug output shows the create tool is called with empty params:
```
[DEBUG] ========================================
[DEBUG] Executing tool: create
[DEBUG] Arguments:
[DEBUG] ========================================
[DEBUG] Tool 'create' completed successfully
[DEBUG] Result preview:
Error creating file: Invalid file_path - must be a non-empty string
```

The `Arguments:` section is empty, meaning `params` is `{}`.

This happens multiple times (3 attempts observed) before the AI gives up.

The AI also tried to call an unknown tool called `implement`:
```
[ERROR] Unknown tool: implement
```

## Notes

- Other tools like `search`, `query`, `extract` work correctly with the same provider
- The issue seems specific to `create` and `edit` tools
- The XML parser (`parseXmlToolCallWithThinking`) may be extracting an empty params object from the AI's response

## Possible Causes

1. The Google model might be outputting the tool call in a different format than expected
2. The XML tool definition for create/edit might not be clear enough for Google models
3. There could be an issue with how Google's response is parsed for these specific tools

## Workaround

None known yet. The tools simply don't work with Google provider.

--- a/npm/src/tools/common.js
+++ b/npm/src/tools/common.js
@@ -5,6 +5,7 @@
 
 import { z } from 'zod';
 import { resolve, isAbsolute } from 'path';
+import { editSchema, createSchema } from './edit.js';
 
 // Common schemas for tool parameters (used for internal execution after XML parsing)
 export const searchSchema = z.object({
@@ -359,14 +360,16 @@ export function buildToolTagPattern(tools = DEFAULT_VALID_TOOLS) {
  * @returns {string[]} - Array of valid parameter names for this tool
  */
 function getValidParamsForTool(toolName) {
-	// Map tool names to their schemas
+	// Map tool names to their schemas (supports both Zod and JSON Schema formats)
 	const schemaMap = {
 		search: searchSchema,
 		query: querySchema,
 		extract: extractSchema,
 		delegate: delegateSchema,
 		bash: bashSchema,
-		attempt_completion: attemptCompletionSchema
+		attempt_completion: attemptCompletionSchema,
+		edit: editSchema,
+		create: createSchema
 	};
 
 	const schema = schemaMap[toolName];
@@ -382,8 +385,13 @@ function getValidParamsForTool(toolName) {
 	}
 
 	// Extract keys from Zod schema
-	if (schema && schema._def && schema._def.shape) {
+	if (schema._def && schema._def.shape) {
 		return Object.keys(schema._def.shape());
+	}
+
+	// Extract keys from JSON Schema (used by edit and create tools)
+	if (schema.properties) {
+		return Object.keys(schema.properties);
 	}
 
 	// Fallback: return empty array if we can't extract schema keys

--- a/npm/tests/unit/xmlParsing.test.js
+++ b/npm/tests/unit/xmlParsing.test.js
@@ -821,4 +821,162 @@ I should search for this...
       });
     });
   });
+
+  describe('Edit and Create tool parsing (Issue #349)', () => {
+    test('should parse edit tool with all parameters', () => {
+      const validTools = ['search', 'edit', 'create', 'attempt_completion'];
+      const aiResponse = `<edit>
+<file_path>src/main.js</file_path>
+<old_string>function oldName() {
+  return 42;
+}</old_string>
+<new_string>function newName() {
+  return 42;
+}</new_string>
+</edit>`;
+
+      const result = parseXmlToolCallWithThinking(aiResponse, validTools);
+
+      expect(result).toEqual({
+        toolName: 'edit',
+        params: {
+          file_path: 'src/main.js',
+          old_string: `function oldName() {
+  return 42;
+}`,
+          new_string: `function newName() {
+  return 42;
+}`
+        }
+      });
+    });
+
+    test('should parse edit tool with replace_all parameter', () => {
+      const validTools = ['edit'];
+      const aiResponse = `<edit>
+<file_path>config.json</file_path>
+<old_string>"debug": false</old_string>
+<new_string>"debug": true</new_string>
+<replace_all>true</replace_all>
+</edit>`;
+
+      const result = parseXmlToolCall(aiResponse, validTools);
+
+      expect(result).toEqual({
+        toolName: 'edit',
+        params: {
+          file_path: 'config.json',
+          old_string: '"debug": false',
+          new_string: '"debug": true',
+          replace_all: true
+        }
+      });
+    });
+
+    test('should parse create tool with all parameters', () => {
+      const validTools = ['create'];
+      const aiResponse = `<create>
+<file_path>src/newFile.js</file_path>
+<content>export function hello() {
+  return "Hello, world!";
+}</content>
+</create>`;
+
+      const result = parseXmlToolCall(aiResponse, validTools);
+
+      expect(result).toEqual({
+        toolName: 'create',
+        params: {
+          file_path: 'src/newFile.js',
+          content: `export function hello() {
+  return "Hello, world!";
+}`
+        }
+      });
+    });
+
+    test('should parse create tool with overwrite parameter', () => {
+      const validTools = ['create'];
+      const aiResponse = `<create>
+<file_path>README.md</file_path>
+<content># My Project
+
+This is a new project.</content>
+<overwrite>true</overwrite>
+</create>`;
+
+      const result = parseXmlToolCall(aiResponse, validTools);
+
+      expect(result).toEqual({
+        toolName: 'create',
+        params: {
+          file_path: 'README.md',
+          content: `# My Project
+
+This is a new project.`,
+          overwrite: true
+        }
+      });
+    });
+
+    test('should handle edit tool with unclosed tags', () => {
+      const validTools = ['edit'];
+      const aiResponse = `<edit>
+<file_path>test.js
+<old_string>foo
+<new_string>bar`;
+
+      const result = parseXmlToolCall(aiResponse, validTools);
+
+      expect(result).toEqual({
+        toolName: 'edit',
+        params: {
+          file_path: 'test.js',
+          old_string: 'foo',
+          new_string: 'bar'
+        }
+      });
+    });
+
+    test('should handle create tool with unclosed tags', () => {
+      const validTools = ['create'];
+      const aiResponse = `<create>
+<file_path>new.txt
+<content>Hello World`;
+
+      const result = parseXmlToolCall(aiResponse, validTools);
+
+      expect(result).toEqual({
+        toolName: 'create',
+        params: {
+          file_path: 'new.txt',
+          content: 'Hello World'
+        }
+      });
+    });
+
+    test('should handle edit tool inside thinking tags', () => {
+      const validTools = ['edit'];
+      const aiResponse = `<thinking>
+I need to rename this function to follow the naming convention.
+</thinking>
+
+<edit>
+<file_path>src/utils.js</file_path>
+<old_string>const getUserName = () => {}</old_string>
+<new_string>const getUsername = () => {}</new_string>
+</edit>`;
+
+      const result = parseXmlToolCallWithThinking(aiResponse, validTools);
+
+      expect(result).toEqual({
+        toolName: 'edit',
+        params: {
+          file_path: 'src/utils.js',
+          old_string: 'const getUserName = () => {}',
+          new_string: 'const getUsername = () => {}'
+        }
+      });
+    });
+  });
 });

